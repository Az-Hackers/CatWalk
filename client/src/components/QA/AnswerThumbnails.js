/* eslint-disable react/prop-types */
import React from 'react';

import Thumbnail from './Thumbnail';

const AnswerThumbnails = ({ thumbnails }) => (
  <div className="QA-AnswerThumbnails">
    {/* {
      thumbnails.map((thumbnail) => (
        <Thumbnail
          thumbnail={thumbnail.url}
          key={thumbnail.id}
        />
      ))
    } */}
  </div>
);

export default AnswerThumbnails;
