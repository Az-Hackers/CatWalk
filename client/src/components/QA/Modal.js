import React, {useState, useEffect} from 'react';
import { Button } from '@material-ui/core';

const api = require('../../../../helpers/qa.js');
const helpers = require('./componentHelpers.js');

const Modal = ({questionModal, question, product, toggleModal, getQuestions, getAnswers}) => {
  const [inputs, setInputs] = useState({questionOrAnswer: '', nickname: 'Example: bobjohnson88', email: ''});
  const [errorMessage, setErrorMessage] = useState('');
  const [errorExists, setErrorExists] = useState(false);
  const [submission, setSubmission] = useState(false);

  useEffect ( () => {
    if (!errorExists && submission) {
      submit();
    } else {
      setSubmission(false);
    }
  }, [submission])

  const submit = () => {
    if (questionModal) {
      const params = {
        body: inputs.questionOrAnswer,
        name: inputs.nickname,
        email: inputs.email,
        product_id: product.id
      };
      api.submitQuestion(params)
        .then(response => getQuestions(product.id));

    } else {
      const params = {
        body: inputs.questionOrAnswer,
        name: inputs.nickname,
        email: inputs.email,
      };
      api.submitAnswer(question.question_id, params)
        .then(response => getAnswers(question.question_id));
    }

    setSubmission(false);
    toggleModal();
  };

  const validate = () => {
    if (!helpers.validateInputs(inputs)) {
      setErrorExists(true);
      setErrorMessage('Please ensure all forms are filled out and that your email is formatted correctly!')
    } else {
      setErrorExists(false);
    }

    if (!helpers.validateEmail(inputs.email)) {
      setErrorExists(true);
      setErrorMessage('Please ensure all forms are filled out and that your email is formatted correctly!');
    } else {
      setErrorExists(false);
    }

    setSubmission(true);
  };

  const handleInputChange = (e) => {
    e.persist();
    const updatedValue = {};
    updatedValue[e.target.name] = e.target.value;
    setInputs({...inputs, ...updatedValue});
  };

  return (
    <div className="QA-Modal">
      <div className="QA-ModalGuts">
      {questionModal ? (<><div>Ask Your Question</div>
        <div>About the {product.name} Here</div></>) : (<><div>Submit Your Answer</div><p>{product.name}: {question.question_body}</p></>)}

      {questionModal ? (<div>Your Question</div>) : (<div>Your Answer</div>)}
        <textarea name="questionOrAnswer" value={inputs.questionOrAnswer} onChange={(event) => handleInputChange(event)}></textarea>

        <div>What is your nickname?</div>
        <input name="nickname" value={inputs.nickname} onChange={(event) => handleInputChange(event)}></input>
        <p>For privacy reasons, do not use your full name or email address</p>

        <div>Your Email</div>
        <input name="email" value={inputs.email} onChange={(event) => handleInputChange(event)}></input>
        <p>For authentication reasons, you will not be emailed</p>

        {errorExists && (<div className="QA-ModalError">{errorMessage}</div>)}

        <Button color="inherit" onClick={(event) => toggleModal(event)}>Cancel</Button>
      {questionModal ? (<Button color="inherit" onClick={() => validate()}>Submit Question</Button>) :
        (<Button color ="inherit" onClick={() => validate()}>Submit Answer</Button>)}
      </div>
    </div>
  );

};

export default Modal;

