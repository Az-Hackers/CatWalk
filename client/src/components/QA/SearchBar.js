/* eslint-disable no-unused-expressions */
/* eslint-disable react/prop-types */
import React, { useState, useEffect } from 'react';
import SearchBar from 'material-ui-search-bar';

const Search = ({ questions, setFilteredQuestions, setCount }) => {
  const [searchInput, setSearchInput] = useState('');

  const filterQuestions = () => {
    (searchInput.length >= 3 && searchInput !== 'Have Questions? Search for answers...')
      ? setFilteredQuestions(questions.filter(({question_body}) => question_body.toLowerCase().includes(searchInput.toLowerCase())))
      : setFilteredQuestions(null);
    setCount(2);
  };

  useEffect(() => {
    filterQuestions();
  }, [searchInput]);

  return (
    <SearchBar
      className="QA-SearchBar"
      placeholder="Have Questions? Search for answers..."
      value={searchInput}
      onChange={(value) => setSearchInput(value)}
      onClick={() => setSearchInput('')}
    />
  );
};

export default Search;
