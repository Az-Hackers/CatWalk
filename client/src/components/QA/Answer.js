import React, {useState} from 'react';
import AnswerThumbnails from './AnswerThumbnails.js';

const api = require('../../../../helpers/qa.js');

const Answer = ({answer}) => {
  const [helpfulness, setHelpfulness] = useState(answer.helpfulness);
  const [helpfulClicked, setHelpfulClicked] = useState(false);
  const [reportClicked, setReportClicked] = useState(false);
  const [reported, setReported] = useState(false);

  const formatDate = (date) => {
    const monthNames = ["January", "February", "March",  "April", "May", "June",  "July", "August", "September", "October", "November", "December"];
    let day = answer.date.slice(8, answer.date.indexOf('T'));
    var month = monthNames[Number(answer.date.slice(6, 7)) - 1];
    var year = answer.date.slice(0, 4);
    var formattedDate = `${month} ${day}, ${year}`;
    return formattedDate;
  };

  const markAnswerHelpful = (e) => {
    e.preventDefault();
    if (helpfulClicked) {
      return;
    } else {
      setHelpfulness(helpfulness + 1);
      setHelpfulClicked(true);
      api.markQuestionOrAnswerHelpful('answers', answer.answer_id);
    };
  };

  const reportAnswer = (e) => {
    e.preventDefault();
    if (reportClicked) {
      return;
    } else {
      setReportClicked(true);
      setReported(true);
      api.reportQuestionOrAnswer('answers', answer.answer_id);
    }
  };

  return (
    <span className="QA-Answer">
    <span className="QA-AnswerTitle medium">{answer.body}</span>
    <p className="QA-AnswerInteractive small">by {answer.answerer_name}, {formatDate(answer.date)} | Helpful? <a href="#" onClick={(event) => markAnswerHelpful(event)}>Yes</a> ({helpfulness}) | &nbsp;
    {!reported ? (<a href="#" onClick={(event) => reportAnswer(event)}>Report</a>) : (<span>Reported</span>)}</p>
    <AnswerThumbnails thumbnails={answer.photos}/>
    </span>

  );

};

export default Answer;