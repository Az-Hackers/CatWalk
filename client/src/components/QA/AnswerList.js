import React from 'react';
import Answer from './Answer.js';

const AnswerList = ({answers}) => (
  <div className="QA-answerList">
  <ul>
      {
        answers.map(answer => {
          return <Answer
          answer={answer}
          key={answer.answer_id}/>
        })
      }
  </ul>
  </div>
);

export default AnswerList;