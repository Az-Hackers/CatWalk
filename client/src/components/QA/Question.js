import React, {useState, useEffect} from 'react';
import AnswerList from './AnswerList.js';
import Modal from './Modal.js';

const api = require('../../../../helpers/api');

const Question = ({question, product}) => {
  const [answers, setAnswers] = useState([]);
  const [count, setCount] = useState(2);
  const [helpfulClicked, setHelpfulClicked] = useState(false);
  const [reportClicked, setReportClicked] = useState(false);
  const [displayModal, setDisplayModal] = useState(false);

  useEffect( () => {
    setCount(2);
    getAnswers();
  }, [question]);

  const getAnswers = () => {
    api.getAnswers(question.question_id, (err, results) => {
      setAnswers(results);
    });
  };

  const markQuestionHelpful = (e) => {
    e.preventDefault();
    if (helpfulClicked) {
      return;
    } else {
      setHelpfulClicked(true);

      api.markQuestionHelpful(question.question_id,
      (err, results) => {
        if (err) {
          console.log(err);
        } else {
          console.log(results);
        }
      });
    };
  };

  const reportQuestion = (e) => {
    e.preventDefault();
    if (reportClicked) {
      return;
    } else {
      setReportClicked(true);

      api.reportQuestion(question.question_id, (err, results) => {
        if (err) {
          console.log(err);
        } else {
          console.log(results);
        }
      })
    };
  };

  const toggleModal = (e) => {
    e.preventDefault();
    setDisplayModal(!displayModal);
  };

  return (
    <div className="QA-question">
    <div>Q: {question.question_body}</div>

    <p>Helpful? <a href="#" onClick={(event) => markQuestionHelpful(event)}>Yes</a> ({question.question_helpfulness}) | <a href="#" onClick={(event) => toggleModal(event)}>Add Answer</a> | <a href="#" onClick={(event) => reportQuestion(event)}>Report</a></p>
    <p>A:</p><AnswerList answers={answers.slice(0, count)}/>
    {answers.length > 2 && (<button className="QA-moreAnswersButton button"
    onClick={() => {
      setCount(count + 2)}}>
      LOAD MORE ANSWERS</button>)}
    {displayModal && (<Modal questionModal={false} question={question.question_body} product={product} displayModal={displayModal} toggleModal={toggleModal}/>)}
    </div>
  );

};

export default Question;