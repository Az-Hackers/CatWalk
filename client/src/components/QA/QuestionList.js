import React from 'react';
import Question from './Question.js';

const QuestionList = ({questions, product}) => (
  <div className="QA-questionList">
  <ul>
      {
        questions.map(question => {
          return <Question
          question={question}
          product={product}
          key={question.question_id}/>
        })
      }
  </ul>
  </div>
);

export default QuestionList;