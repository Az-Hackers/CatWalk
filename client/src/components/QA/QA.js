import React, {useState, useEffect} from 'react';
import QuestionList from './QuestionList.js';

const api = require('../../../../helpers/api');

const QA = ({id}) => {
  const [questions, setQuestions] = useState([]);
  const [count, setCount] = useState(4);

  useEffect( () => {
    console.log('idChange useEffect');
    setCount(4);
    getQuestions();
  }, [id])

  useEffect( () => {
    getQuestions();
  }, [count])

  useEffect( () => {
    if(questions.length > count) {
      setQuestions(questions => questions = questions.slice(0, count));
      console.log('count did not update in time, but I got you fam!');
    }
  }, [questions.length])

  const getQuestions = () => {
    api.getQuestions(id, count, (err, results) => {
      setQuestions(results);
    })
  };

  return (
    <div className="QA">
    <div>QUESTIONS and ANSWERS</div>
    <div>Search Bar Placeholder</div>
    <QuestionList questions={questions} />
    {questions.length > 2 && (<button onClick={() => {
      setCount(count + 2)}}>
        MORE ANSWERED QUESTIONS</button>)}
    <button>ADD A QUESTION +</button>
    </div>
  )
};

export default QA;