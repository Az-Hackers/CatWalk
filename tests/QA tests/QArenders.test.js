import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import QA from '../../client/src/components/QA/QA.js';
import QuestionList from '../../client/src/components/QA/QuestionList.js';
import Question from '../../client/src/components/QA/Question.js';
import AnswerList from '../../client/src/components/QA/AnswerList.js';
import Answer from '../../client/src/components/QA/Answer.js';
import Modal from '../../client/src/components/QA/Modal.js';
import SearchBar from '../../client/src/components/QA/SearchBar.js';
import Thumbnail from '../../client/src/components/QA/Thumbnail.js';
import AnswerThumbnails from '../../client/src/components/QA/AnswerThumbnails.js';

//shallow render tests for QA components

    describe("QA components should all successfully render in isolation", () =>{
      it("should render parent QA component", () => {
        const wrapper = shallow(<QA />);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

      it('parent QA component should render SearchBar', () => {
        const wrapper = shallow(<QA />);
        expect(wrapper.find(SearchBar)).toHaveLength(1);
      });

      it("should render QuestionList component", () => {
        const questions = [{question_id: 1, question_body: 'et tu Brutu'}];
        const wrapper = shallow(<QuestionList questions={questions} product={5}/>);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

      it("should render Question component", () => {
        const question = {question_body: 'Hello world'};
        const wrapper = shallow(<Question question={question}/>);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it("should render AnswerList component", () => {
      const answers = [{placeholder: 'a', answer_id: 1}, {placeholder: 'a', answer_id: 2}, {placeholder: 'a', answer_id: 3}];
      const wrapper = shallow(<AnswerList answers={answers}/>);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it("should render Answer component", () => {
      const answer = {body: 'hello world', answerername: 'bob', answer_id: 5, helpfulness: 10, photos: [], date: '2018-08-18T00:00:00.000Z'};
      const wrapper = shallow(<Answer answer={answer}/>);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it("should render Modal component", () => {
      const product = {name: 'Gucci shoes', id: 1};
      const wrapper = shallow(<Modal product={product} questionModal={true}/>);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it("should render SearchBar component", () => {
      const wrapper = shallow(<SearchBar />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it("should render Thumbnail component", () => {
      const wrapper = shallow(<Thumbnail />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it("should render AnswerThumbnails component", () => {
      const thumbnails = [{url: "urlplaceholder/answer_5_photo_number_1.jpg", id: 1}]
      const wrapper = shallow(<AnswerThumbnails thumbnails={thumbnails}/>);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
});