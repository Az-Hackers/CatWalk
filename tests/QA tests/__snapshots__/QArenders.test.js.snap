// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QA components should all successfully render in isolation should render Answer component 1`] = `
<span
  className="QA-Answer"
>
  <span
    className="QA-AnswerTitle medium"
  >
    hello world
  </span>
  <AnswerThumbnails
    thumbnails={Array []}
  />
  <p
    className="QA-AnswerInteractive small"
  >
    by 
    , 
    August 18, 2018
     | Helpful? 
    <a
      href="#"
      onClick={[Function]}
    >
      Yes
    </a>
     (
    10
    ) |  
    <a
      href="#"
      onClick={[Function]}
    >
      Report
    </a>
  </p>
</span>
`;

exports[`QA components should all successfully render in isolation should render AnswerList component 1`] = `
<div
  className="QA-AnswerList"
>
  <span
    className="QA-AnswerListAnchor medium"
  >
    A:
  </span>
  <ul
    className="QA-Answers"
  >
    <Answer
      answer={
        Object {
          "answer_id": 1,
          "placeholder": "a",
        }
      }
      key="1"
    />
    <Answer
      answer={
        Object {
          "answer_id": 2,
          "placeholder": "a",
        }
      }
      key="2"
    />
  </ul>
  <WithStyles(ForwardRef(Button))
    className="QA-moreAnswersButton small"
    color="primary"
    onClick={[Function]}
    size="small"
  >
    LOAD MORE ANSWERS
  </WithStyles(ForwardRef(Button))>
</div>
`;

exports[`QA components should all successfully render in isolation should render AnswerThumbnails component 1`] = `
<div
  className="AnswerThumbnails"
>
  <Thumbnail
    key="1"
    thumbnail="urlplaceholder/answer_5_photo_number_1.jpg"
  />
</div>
`;

exports[`QA components should all successfully render in isolation should render Modal component 1`] = `
<div
  className="QA-Modal"
>
  <div
    className="QA-ModalGuts"
  >
    <div>
      Ask Your Question
    </div>
    <div>
      About the 
      Gucci shoes
       Here
    </div>
    <div>
      Your Question
    </div>
    <textarea
      name="questionOrAnswer"
      onChange={[Function]}
      value=""
    />
    <div>
      What is your nickname?
    </div>
    <input
      name="nickname"
      onChange={[Function]}
      value="Example: bobjohnson88"
    />
    <p>
      For privacy reasons, do not use your full name or email address
    </p>
    <div>
      Your Email
    </div>
    <input
      name="email"
      onChange={[Function]}
      value=""
    />
    <p>
      For authentication reasons, you will not be emailed
    </p>
    <WithStyles(ForwardRef(Button))
      color="inherit"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="inherit"
      onClick={[Function]}
    >
      Submit Question
    </WithStyles(ForwardRef(Button))>
  </div>
</div>
`;

exports[`QA components should all successfully render in isolation should render Question component 1`] = `
<div
  className="QA-Question"
>
  <span
    className="Question-Title medium"
  >
    Q:  
    Hello world
  </span>
  <p
    className="Question-Interactive small"
  >
    Helpful? 
    <a
      href="#"
      onClick={[Function]}
    >
      Yes
    </a>
     (
    ) | 
    <a
      href="#"
      onClick={[Function]}
    >
      Add Answer
    </a>
  </p>
  <AnswerList
    answers={Array []}
    setAnswers={[Function]}
  />
</div>
`;

exports[`QA components should all successfully render in isolation should render QuestionList component 1`] = `
<Fragment>
  <ul
    className="QA-QuestionList"
  >
    <Question
      key="1"
      product={5}
      question={
        Object {
          "question_body": "et tu Brutu",
          "question_id": 1,
        }
      }
    />
  </ul>
</Fragment>
`;

exports[`QA components should all successfully render in isolation should render SearchBar component 1`] = `
<WithStyles(ForwardRef)
  className="QA-SearchBar"
  onChange={[Function]}
  onClick={[Function]}
  style={
    Object {
      "color": "rgba(77,100,154,1)",
    }
  }
  value="Have Questions? Search for answers..."
/>
`;

exports[`QA components should all successfully render in isolation should render Thumbnail component 1`] = `
<Fragment>
  <img
    alt="no image"
    className="AnswerThumbnail"
    onClick={[Function]}
  />
</Fragment>
`;

exports[`QA components should all successfully render in isolation should render parent QA component 1`] = `
<div>
  <QA
    onClick={[Function]}
  />
</div>
`;
